source/utils/loop_iterator.hpp

```diff
@@ -0,0 +1,44 @@
+#pragma once
+
+#include <iterator>
+#include <stdexcept>
+
+namespace sim {
+
+template <std::forward_iterator Iter>
+class LoopIterator {
+public:
+    using value_t = typename std::iterator_traits<Iter>::value_type;
+    using value_ref_t = typename std::iterator_traits<Iter>::reference;
+
+    LoopIterator() = default;
+
+    LoopIterator(Iter a_begin, Iter a_end)
+        : m_curr(a_begin), m_begin(a_begin), m_end(a_end) {
+        if (a_begin == a_end) {
+            throw std::runtime_error("LoopIterator range cannot be empty");
+        }
+    }
```

Comment:
Let's change logic the extreme case `m_begin = m_end`. Let allow to construct such iterator, but in `operator++` of it just white error message to log and return `m_end`, in `operator*` write critical message to log and return dereferenced (and potentially invalid) value
```suggestion
    LoopIterator(Iter a_begin, Iter a_end)
        : m_curr(a_begin), m_begin(a_begin), m_end(a_end) {}
```

Reply:
> Let's change logic the extreme case `m_begin = m_end`. Let allow to construct such iterator, but in `operator++` of it just white error message to log and return `m_end`, in `operator*` write critical message to log and return dereferenced (and potentially invalid) value
> 
> To simply use iterator, in `next_[in\out]link` instead of check if the link container is empty, check that corresponding iterator is equal to the end of container

In the beginning of problem discussion i wrote you to throw and error in incorrect iterator. But later in discussion with @ArtyomPeshkov we found better solution (suggested now). Sorry for this vagueness and spending your time